design/backup-performance-improvements.md

```diff
@@ -0,0 +1,212 @@
+The first goal is to enhance backup performance by allowing the primary backup controller to run in multiple threads, enabling Velero to back up multiple items at the same time for a given backup.
+The second goal is to enable Velero to eventually support VolumeGroupSnapshots.
+For both of these goals, Velero needs a way to determine which items should be backed up together.
+
+This design proposal will include two development phases:
+- Phase 1 will refactor the backup workflow to identify blocks of items that should be backed up together, and then coordinate backup hooks among items in the block.
+- Phase 2 will add multiple multiple worker threads for backing up item blocks, so instead of backing up each block as it identified, the velero backup workflow will instead add the block to a channel and one of the workers will pick it up.
+- Actual support for VolumeGroupSnapshots is out-of-scope here and will be handled in a future design proposal, but the item block refactor introduced in Phase 1 is a primary building block for this future proposal.
+
+## Background
+Currently, during backup processing, the main Velero backup controller runs in a single thread, completely finishing the primary backup processing for one resource before moving on to the next one.
+We can improve the overall backup performance by backing up multiple items for a backup at the same time, but before we can do this we must first identify resources that need to be backed up together.
+As part of this initial refactoring, once these "Item Blocks" are identified, an additional change will be to move pod hook processing up to the ItemBlock level.
+If there are multiple pods in the ItemBlock, pre-hooks for all pods will be run before backing up the items, followed by post-hooks for all pods.
+This change to hook processing is another prerequisite for future VolumeGroupSnapshot support, since supporting this will require backing up the pods and volumes together for any volumes which belong to the same group.
+Once we are backing up items by block, the next step will be to create multiple worker threads to process and back up ItemBlocks, so that we can back up multiple ItemBlocks at the same time.
+
+## Goals
+- Identify groups of items to back up together (ItemBlocks).
+- Manage backup hooks at the ItemBlock level rather than per-item.
+- Using worker threads, back up ItemBlocks at the same time.
+
+## Non Goals
+- Support VolumeGroupSnapshots: this is a future feature, although certain prerequisites for this enhancement are included in this proposal.
+- Process multiple backups in parallel: this is a future feature, although certain prerequisites for this enhancement are included in this proposal.
+- Refactoring plugin infrastructure to avoid RPC calls for internal plugins.
+
+## High-Level Design
+
+### Phase 1: ItemBlock processing
+- A new BIA method, `GetAdditionalItems`, will be needed for pre-processing ItemBlocks (this will require a new BIAv3 API).
```

Comment:
That is addressed in the detailed design. The one case where we cannot predict is when we create new items (i.e. VolumeSnapshot created by PVC plugin). In this case, it is expected that the new API call will not return the not-yet-created item. This doesn't cause a problem, though, since because the item doesn't exist yet, it won't be in the  Item Collector list, so it won't be in any of the ItemBlocks. It gets backed up just like today, when Execute returns the AdditionalItem.

For existing BIAs, the way to handle this is also called out in detailed design. The short answer is that in order to back up items in parallel, all BIAs registered must implement v3 -- i.e. v2 BIAs won't be adaptable to v3. For BIAs supported by Velero (including CSI, which is now in-tree), those will all need to be upgraded to V3 as part of the Phase 1 implementation.

Because it's a breaking change, we have a new BIA v3 version. If users have custom BIAs (for example, the openshift plugin in OADP) that have not been upgraded, then the v3 plugins internal to Velero will be adapted to v2 and Velero will respond by limiting the worker pool to a size of 1, so we get the current one-item-at-a-time sequential backup.

In other words, if a user adds a v1 or v2 plugin, Velero degrades gracefully to current behavior, probably with a warning logged that they should upgrade their plugins if they want to make use of the new enhanced functionality. Once they upgrade their plugins, they get the new functionality.

Reply:
>  it is expected that the new API call will not return the not-yet-created item

Take PVC plugin creating VolumeSnapshot for example, let say we are backing up 3 PVCs, all of them should be in one group according to upper level/user rules, then for `Execute` of each PVC, a VolumeSnapshot is returned so BIA further handle this VolumeSnapshot. As a result, it will still be in sequence PVC1->VolumeSnapshot1->PVC2->VolumeSnapshot2->PVC3->VolumeSnapshot3. There is no parallelism.